packages/gatsby/src/query/query-compiler.js

```diff
@@ -435,7 +435,9 @@ const determineUsedFragmentsForDefinition = (
 ) => {
   const { def, name, isFragment, filePath } = definition
   const cachedUsedFragments = fragmentsUsedByFragment.get(name)
-  if (cachedUsedFragments) {
+
+  // `cachedUsedFragments` could be a Set with size 0 which happens to be truthy
```

Comment:
Note that Set with 0 size can be correct (i.e. when fragment doesn't actually use other fragments).

The problem here is sometimes it's 0 when it shouldn't be

So this adds additional works sometimes, but it's easiest hot fix we can come up without diving too much into this code, which will require @vladar expertise most likely